### ISSUE TITLE
ipv6 not working in SoftAP mode (IDFGH-3913)

### ISSUE THREAD

**Issue by fariouche** [espressif/esp-idf#5807]
## Environment

- Development Kit:      ESP32-Wrover-Kit
- Kit version (for WroverKit/PicoKit/DevKitC): [v4]
- Module or chip used:  [ESP32-WROOM-32]
- IDF version : "v4.3-dev-907-g6c17e3a64" (git master 6c17e3a64)
- Build System:         [Make]
- Compiler version:  xtensa-esp32-elf-gcc (crosstool-NG esp-2020r2) 8.2.0
- Operating System:     [Linux]
- Using an IDE?: [No]
- Power Supply:         [USB]

## Problem Description
I'm building a ipv6 & ipv4 softap.
Using the simple softap example, I can run the softap, connect my smartphone and get an ip address. Good!
My issue is with ipv6... 
My code waits for the WIFI_EVENT_AP_START event, and when I get it, I call esp_netif_create_ip6_linklocal()
I see the netif debug trace saying "netif: IPv6 address 0 of interface  set to FE80::200:FF:FE00:0/0x08"
But I never receive the IP_EVENT_GOT_IP6.
If I try to call esp_netif_get_ip6_linklocal() after WIFI_EVENT_AP_START, I get an error -1.

Further investigations shows that the failure reason is because of the ip6 address not being PREFERED.

IPv6 works fine in station mode.

I'm not sure where to look at, and I unfortunately really need ipv6.

### Steps to reproduce
Just compile the softap_example_main.c example with the following additions:
ESP_ERROR_CHECK(esp_event_handler_instance_register(IP_EVENT,
                                                        ESP_EVENT_ANY_ID,
                                                        &ip_event_handler,
                                                        NULL,
                                                        NULL));
just after the already existing call to esp_event_handler_instance_register.

The ip_event_handler function:

static void ip_event_handler(void* arg, esp_event_base_t event_base,
                                    int32_t event_id, void* event_data)
{
 
    switch (event_id) 
    {
    case IP_EVENT_STA_GOT_IP:
        ESP_LOGI(GATTC_TAG, "got ipv4");
    break;
        
    case IP_EVENT_GOT_IP6:
    {
        ESP_LOGI(GATTC_TAG, "got ipv6"); 
    }
    break;

    default:
        break;
    }
}

and inside the wifi_event_handler:
if(event_id == WIFI_EVENT_AP_START)
{
        esp_err_t err;
        ESP_LOGI(GATTC_TAG, "WIFI_EVENT_AP_START");
       
        if(!esp_netif_is_netif_up(g_sta_netif))
            ESP_LOGE(GATTC_TAG, "netif not up!");
        err = esp_netif_create_ip6_linklocal(g_sta_netif);
        if(err != ESP_OK)
            ESP_LOGE(GATTC_TAG, "esp_netif_create_ip6_linklocal failed, %d", err);
        
        err = esp_netif_get_ip6_linklocal(g_sta_netif, (esp_ip6_addr_t*)&if_ipaddr_ip6);
        if ( err != ESP_OK) {
            ESP_LOGE(TAG, "get IPv6 address failed %d", err);
        } else {
            static char str[IP6ADDR_STRLEN_MAX];
            ip6addr_ntoa_r(&if_ipaddr_ip6, str, IP6ADDR_STRLEN_MAX);
            ESP_LOGI(TAG, "got IPv6 addr:%s", str);
        }
 }
   

## Debug Logs
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4
load:0x3fff0034,len:7152
ho 0 tail 12 room 4
load:0x40078000,len:13836
load:0x40080400,len:3948
entry 0x40080674
I (30) boot: ESP-IDF v4.3-dev-907-g6c17e3a64-dirty 2nd stage bootloader
I (31) boot: compile time 14:07:54
I (32) boot: chip revision: 1
I (35) boot_comm: chip revision: 1, min. bootloader chip revision: 0
I (42) boot.esp32: SPI Speed      : 40MHz
I (47) boot.esp32: SPI Mode       : DIO
I (52) boot.esp32: SPI Flash Size : 4MB
I (56) boot: Enabling RNG early entropy source...
I (62) boot: Partition Table:
I (65) boot: ## Label            Usage          Type ST Offset   Length
I (72) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (80) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (87) boot:  2 ota_0            OTA app          00 10 00010000 001d0000
I (95) boot:  3 ota_1            OTA app          00 11 001e0000 001d0000
I (102) boot:  4 creds            Unknown data     01 82 003b0000 00020000
I (110) boot:  5 devdata          Unknown data     01 82 003d0000 00020000
I (117) boot:  6 settings         Unknown data     01 82 003f0000 00010000
I (125) boot: End of partition table
I (129) boot_comm: chip revision: 1, min. application chip revision: 0
I (136) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x3cd34 (249140) map
I (240) esp_image: segment 1: paddr=0x0004cd5c vaddr=0x3ffbdb60 size=0x032bc ( 12988) load
I (246) esp_image: segment 2: paddr=0x00050020 vaddr=0x400d0020 size=0xf88b8 (1018040) map
I (635) esp_image: segment 3: paddr=0x001488e0 vaddr=0x3ffc0e1c size=0x01cb4 (  7348) load
I (638) esp_image: segment 4: paddr=0x0014a59c vaddr=0x40080000 size=0x00400 (  1024) load
I (642) esp_image: segment 5: paddr=0x0014a9a4 vaddr=0x40080400 size=0x1d060 (118880) load
I (718) boot: Loaded app from partition at offset 0x10000
I (718) boot: Disabling RNG early entropy source...
I (729) cpu_start: cpu freq: 160
I (729) cpu_start: Pro cpu up.
I (729) cpu_start: Starting app cpu, entry point is 0x400810c4
I (0) cpu_start: App cpu up.
I (750) cpu_start: Pro cpu start user code
I (750) cpu_start: Application information:
I (750) cpu_start: Project name:     careos_box
I (755) cpu_start: App version:      0.95
I (760) cpu_start: Compile time:     Aug 28 2020 14:07:53
I (766) cpu_start: ELF file SHA256:  1432c3aedbc38230...
I (772) cpu_start: ESP-IDF:          v4.3-dev-907-g6c17e3a64-dirty
I (779) heap_init: Initializing. RAM available for dynamic allocation:
I (786) heap_init: At 3FFAFF10 len 000000F0 (0 KiB): DRAM
I (792) heap_init: At 3FFB6388 len 00001C78 (7 KiB): DRAM
I (798) heap_init: At 3FFB9A20 len 00004108 (16 KiB): DRAM
I (804) heap_init: At 3FFCAE60 len 000151A0 (84 KiB): DRAM
I (810) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (817) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (823) heap_init: At 4009D460 len 00002BA0 (10 KiB): IRAM
I (830) spi_flash: detected chip: generic
I (834) spi_flash: flash io: dio
I (839) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.

I (130) main server: Partition size: total: 113201, used: 2761
I (150) main server: Partition size: total: 113201, used: 0
I (150) main server: Partition size: total: 52961, used: 0
I (160) BTDM_INIT: BT controller compile version [270f54d]
I (1172) system_api: Base MAC address is not set
I (1182) system_api: read default base MAC address from EFUSE
W (1192) phy_init: failed to load RF calibration data (0xffffffff), falling back to full calibration
I (1342) phy: phy_version: 4370, 4e803b3, Aug 11 2020, 14:18:07, 0, 2
netif: netmask of interface  set to 255.0.0.0
netif: GW address of interface  set to 127.0.0.1
netif_set_ipaddr: netif address being changed
I (1702) BT_GATT: GATT_Register
netif: added interface lo IP addr 127.0.0.1 netmask 255.0.0.0 gw 127.0.0.1
D (1712) BT_GATT: UUID=[0x0000]
I (1722) BT_GATT: allocated gatt_if=1
I (1732) BT_GATT: GATT_Register
D (1732) BT_GATT: UUID=[0x0001]
I (1732) wifi:I (1732) BT_GATT: allocated gatt_if=2
wifi driver task: 3ffdb808, prio:23, stack:6656, core=0
I (1742) BT_GATT: GATT_Register
I (1762) wifi:wifi firmware version: 8d94a49
I (1762) wifi:wifi certification version: v7.0
I (1772) wifi:config NVS flash: enabled
I (1772) wifi:config nano formating: disabled
I (1772) wifi:Init data frame dynamic rx buffer num: 32
I (1782) wifi:Init management frame dynamic rx buffer num: 32
I (1782) wifi:Init management short buffer num: 32
I (1792) wifi:Init dynamic tx buffer num: 32
I (1792) wifi:Init static rx buffer size: 1600
I (1802) wifi:Init static rx buffer num: 10
I (1802) wifi:Init dynamic rx buffer num: 32
I (1802) wifi_init: rx ba win: 6
D (1802) BT_GATT: UUID=[0x0002]
I (1812) BT_GATT: allocated gatt_if=3
I (1812) wifi_init: tcpip mbox: 32
I (1822) BT_GATT: GATT_StartIf gatt_if=1
D (1822) BT_GATT: gatt_find_the_connected_bda start_idx=0
I (1822) wifi_init: udp mbox: 6
I (1832) BT_GATT: GATT_StartIf gatt_if=2
D (1842) BT_GATT: gatt_find_the_connected_bda start_idx=0
I (1832) wifi_init: tcp mbox: 6
I (1842) BT_GATT: GATT_StartIf gatt_if=3
D (1852) BT_GATT: gatt_find_the_connected_bda start_idx=0
I (1852) wifi_init: tcp tx win: 5744
I (1862) wifi_init: tcp rx win: 5744
I (1872) wifi_init: tcp mss: 1440
I (1882) wifi_init: WiFi IRAM OP enabled
I (1892) wifi_init: WiFi RX IRAM OP enabled
---- creating softAP ------
netif: netmask of interface  set to 255.255.255.0
netif: GW address of interface  set to 192.168.4.1
netif_set_ipaddr: netif address being changed
netif: setting default interface 
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 0 of interface  set to FE80::200:FF:FE00:0/0x08
I (1952) wifi:mode : softAP (24:62:ab:b5:a8:89)
I (1952) wifi:Total power save buffer number: 16
I (1952) wifi:Init max length of beacon: 752/752
I (1952) wifi:Init max length of beacon: 752/752
WIFI_EVENT_AP_START
netif: netmask of interface  set to 255.255.255.0
netif: GW address of interface  set to 192.168.4.1
E (1972) main server: get IPv6 address failed -1
netif_set_ipaddr: netif address being changed
netif: added interface ap IP addr 192.168.4.1 netmask 255.255.255.0 gw 192.168.4.1
netif: setting default interface ap
E (3992) main server: get IPv6 address failed -1

**Comment by david-cermak** [comment]
Hi @fariouche 

I can see that the LL address state is 0x08 immediately after calling the `esp_netif_create_ip6_linklocal()` (which is `IP6_ADDR_TENTATIVE`, looking for possibly duplicates on the network).
Could you please check what state the LL address is after 2-5 seconds after the `esp_netif_create_ip6_linklocal()`?
Please insert the below debugging message
```
    ESP_LOGI(TAG, "%s netif_ip6_addr_state=%d", __func__, netif_ip6_addr_state(p_netif, 0));
```
into [`esp_netif_get_ip6_linklocal()`](https://github.com/espressif/esp-idf/blob/master/components/esp_netif/lwip/esp_netif_lwip.c#L1534) and call it multiple times after creating the IPv6 LL?

**Comment by fariouche** [comment]
Hi

I've already done the test, and it was always returning an error. I've redone the test with the debug trace you asked to be sure, calling the function every 2s. Still in TENTATIVE after 20s...

By the way, I tried also to set the lwip flag to not do the tentative retry (LWIP_IPV6_DUP_DETECT_ATTEMPTS), and the same error: (the only difference is that now the trace that displays the ipv6 address shows /0x30 (which means IP6_ADDR_PREFERRED I think)

netif: netmask of interface  set to 255.255.255.0
netif: GW address of interface  set to 192.168.4.1
netif_set_ipaddr: netif address being changed
netif: setting default interface 
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 0 of interface  set to FE80::200:FF:FE00:0/0x30
I (1742) wifi:mode : softAP (24:62:ab:b5:a8:89)
I (1742) wifi:Total power save buffer number: 16
I (1742) wifi:Init max length of beacon: 752/752
I (1742) wifi:Init max length of beacon: 752/752
V () careos core init iotivity
I (1742) CAREOS_BLE: WIFI_EVENT_AP_START
V () careos core init done
netif: netmask of interface  set to 255.255.255.0
netif: GW address of interface  set to 192.168.4.1
I (1762) main server: got IPv4 addr:192.168.4.1
netif_set_ipaddr: netif address being changed
netif: added interface ap IP addr 192.168.4.1 netmask 255.255.255.0 gw 192.168.4.1
I (1782) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
netif: setting default interface ap
E (1792) main server: get IPv6 address failed -1
I (3802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
E (3802) main server: get IPv6 address failed -1
I (5802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
E (5802) main server: get IPv6 address failed -1
I (7802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
E (7802) main server: get IPv6 address failed -1
I (9802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
E (9802) main server: get IPv6 address failed -1
I (11802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
E (11802) main server: get IPv6 address failed -1
I (13802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0
E (13802) main server: get IPv6 address failed -1
I (15802) esp_netif_lwip: esp_netif_get_ip6_linklocal netif_ip6_addr_state=0

**Comment by david-cermak** [comment]
Thanks for the update. Unfortunately I am not able to reproduce the issue, could you please also share your `sdkconfig`? (Could think of some configuration options, but AFAIK this state machine should work for all possible settings...)

> I tried also to set the lwip flag to not do the tentative retry

In that case you could use the assigned LL address as is, it should be valid and should work. But the `IP_EVENT_GOT_IP6` wouldn't come as the [`nd6 callback `](https://github.com/espressif/esp-lwip/blob/2.1.2-esp/src/core/ipv6/nd6.c#L1146) gets called only on status change and in this configuration it's directly set to 0x30 (`IP6_ADDR_PREFERRED`)

**Comment by fariouche** [comment]
I see, this is good news if you do not reproduce it. It will be easier to pinpoint the issue.
Do you think you can share the test code you used?
This is the config file I'm using:
[sdkconfig.txt](https://github.com/espressif/esp-idf/files/5151990/sdkconfig.txt)

**Comment by david-cermak** [comment]
@fariouche Thanks for the update, will check your configuration.

I simply placed these lines at the end of the `main()` function:
``` C
    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_GOT_IP6, &on_got_ipv6, NULL));
    vTaskDelay(10000/ portTICK_PERIOD_MS);
    esp_err_t err = esp_netif_create_ip6_linklocal(g_sta_netif);
    ESP_LOGI(TAG, "esp_netif_create_ip6_linklocal returned %d", err);
```
(just assumed the AP to start prior to these 10s, of course adding an event flag, or calling it directly from the handler is a better option)

Plus an IPv6 handler which just prints out the address:
``` c
static void on_got_ipv6(void *arg, esp_event_base_t event_base, int32_t event_id, void *event_data)
{
    ip_event_got_ip6_t *event = (ip_event_got_ip6_t *)event_data;
    ESP_LOGI(TAG, "Got IPv6 event: Interface \"%s\" address: " IPV6STR, esp_netif_get_desc(event->esp_netif),IPV62STR(event->ip6_info.ip));
}
```

**Comment by david-cermak** [comment]
@fariouche I've tested your configuration with the IDF version of 6c17e3a and it works okay for me:
```
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 0 of interface ap set to FE80:3XXXX9/0x08
I (10862) wifi softAP: esp_netif_create_ip6_linklocal returned 0
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 0 of interface ap set to FE80::3XXXX9/0x09
netif_ip6_addr_set_state: netif address state being changed
netif: IPv6 address 0 of interface ap set to FE80::3XXXX9/0x30
I (12602) wifi softAP: Got IPv6 event: Interface "ap" address: fe80:0000:0000:0000:3xxx:xxx:xxx:xxx9
```

Any news from your end? Have you tried the snippets above?

**Comment by fariouche** [comment]
Hello,

Sorry, I'm facing other issues currently that I need to sort out before doing again the investigation. Hopefully soon

**Comment by fariouche** [comment]
Sorry, closed by mistake!
I'm back on the investigation of this issue

I tried your 4 lines of code in the main and guess what, it worked.

The create_linklocal code in my case was done inside the AP_START event and in that case it does not work.

**Comment by fariouche** [comment]
The question now if where should I put the esp_netif_create_ip6_linklocal() call if I cannot in AP_START event?

**Comment by fariouche** [comment]
For information, doing the create_ip6_linklocal 2s after the AP_START event works (the 2s must be asynchronous, so no task_delay inside the AP_START!) I suspect that something is done after the AP_START event returns.

**Comment by david-cermak** [comment]
@fariouche I'd suggest waiting until the related netif is up:
``` c
    ESP_LOGI(TAG, "ESP_WIFI_MODE_AP");
    wifi_init_softap();
    while (!esp_netif_is_netif_up(g_ap_netif)) {
        vTaskDelay(100/ portTICK_PERIOD_MS);
    }
    esp_err_t err = esp_netif_create_ip6_linklocal(g_ap_netif);
    ESP_LOGI(TAG, "esp_netif_create_ip6_linklocal returned %d", err);
```
Unfortunately there is no event to notify about certain netif going up/down.

I might work in the handler of the `WIFI_EVENT_AP_START` event, too, but need to register the handler *after* wifi system handler gets registered in `esp_netif_create_default_wifi_ap()`. Then when the event comes it first starts the netif, queues it into the lwip task, and after that we create local link address again enqueueing into lwip, then the lwip task pre-empts the main task setting the netif up first and then creating local link IPv6 address.
I'd still suggest the wait loop above in the main task as it's I think easier, more readable, robust and less error prone.

**Comment by fariouche** [comment]
Thank you

Blocking the WIFI_EVENT_AP_START will not work (if I do the vtask_delay inside, the netif is never up)
So I suspect that the netif is set up in the same task that handle the AP_START event.

The wait loop you suggest may work, but in my case I will have to create a task just for that to not block the app.

It would have been nice to have a NETIF_UP event (there is one for the ethernet interface)

**Comment by david-cermak** [comment]
> So I suspect that the netif is set up in the same task that handle the AP_START event.
It's not executed in the same task, but queues both action to the same time, which has the same effect. Here however, we could control the order of the handlers, so having `esp_netif_create_default_wifi_ap()` run before `esp_event_handler_instance_register(..., esp_event_handler_instance_register, ..)` should work, even with adding the `vTaskDelay()` directly into the handler (before creating local ink address). This would be the same as the wait loop, but would effectively **block** the entire system not only the app.

> It would have been nice to have a NETIF_UP event (there is one for the ethernet interface)
Do you mean `ETHERNET_EVENT_CONNECTED` event?

I general i think it would be nice to have such an event.

**Comment by fariouche** [comment]
For the vtask_delay in the event handler, I've actually already tried it (put 10s) and it didn't work.
For now I'm using my timer task to queue a timed event to myself when I receive the event AP_START (1s seems to work)

For the ETHER_EVENT_CONNECTED, for netif, a possible solution is maybe fire up the AP_START only after the netif is also up.

This will solve the issue I believe (and maybe raise regressions?)

**Comment by Alvin1Zhang** [comment]
Thanks for reporting and sharing updates, will close the ticket now, feel free to reopen.
